### ISSUE TITLE
FreeBSD: could not start rsync command on remote host

### ISSUE THREAD

**Issue by mqudsi** [microsoft/VSLinux#348]
Is there some way to enable command-level debug logging where each command executed remotely over ssh on the remote build server and the output/error is logged?

Before I stumbled across the fact that RHEL/Fedora remote hosts seem to work problem-free, I was experimenting with different remote build servers and tried out FreeBSD (which is my personal favorite *nix in terms of stability, sanity, simplicity, and reliability) and experienced the same error reported in #341, but this time from the very start (i.e. rsync was never started).

rsync is installed on the remote host, but I imagine the problem is with the detection method you use to find free ports not supporting FreeBSD at all in the first place. There is on `/proc/net/tcp` under TCP (all of `/proc` is a Linux-specific feature, though FreeBSD mimics some of it for Linux compatibility). 

Under FreeBSD, the command `sockstat` can be used with the `-l` parameter to see a list of all active (in-use) listening ports, producing output like this:

```
mqudsi@freebsd ~> sockstat -l
USER     COMMAND    PID   FD PROTO  LOCAL ADDRESS         FOREIGN ADDRESS
www-data nginx      79990 6  tcp4   *:81                  *:*
www-data nginx      79990 7  tcp4   *:82                  *:*
www-data nginx      79990 8  tcp4   *:443                 *:*
www-data nginx      79990 9  tcp4   *:80                  *:*
root     smbd       75251 7  dgram  /var/db/samba4/private/msg.sock/75251
root     smbd       74116 7  dgram  /var/db/samba4/private/msg.sock/74116
root     smbd       12104 7  dgram  /var/db/samba4/private/msg.sock/12104
root     smbd       7433  7  dgram  /var/db/samba4/private/msg.sock/7433
root     java       37216 222 tcp4  *:8080                *:*
root     java       37216 223 stream(not connected)
root     java       37216 231 tcp4  127.0.0.1:8005        *:*
root     java       37216 334 stream(not connected)
root     mountd     1136  6  udp6   *:959                 *:*
root     mountd     1136  7  tcp6   *:959                 *:*
root     mountd     1136  8  udp4   *:959                 *:*
root     mountd     1136  9  tcp4   *:959                 *:*
root     node       1042  11 tcp46  *:25                  *:*
git      gitea      1004  12 tcp46  *:2222                *:*
git      gitea      1004  15 tcp46  *:3000                *:*
...
```

The following sh script (note: bash is NOT installed on FreeBSD machines by default) will return the first free port over 4444:

```sh
#!/bin/sh
# Returns the first free port over 4444
# Author: Mahmoud Al-Qudsi <mqudsi@neosmart.net>
# Copyright NeoSmart Technologies. Released under the MIT License.

IN_USE=$(sockstat -l | tr -s ' ' ' ' | cut -f6 -d ' ' | sed -r 's/^[^:]+//;/::|(:.*){2,}/d;s/[^0-9]+//g;/^$/d' | sort -un)

for p in $(seq 4444 6500)
do
	if ! echo $IN_USE | grep >/dev/null ^$p\$; then
		echo $p
		break
	fi
done
```

**Comment by elizabethmorrow** [comment]
Hi @mqudsi 

We recently added logging of the commands being executed over the ssh shell. To turn this on, go to Tools->Options->Cross Platform -> Logging:
1) Select Enable Logging
2) Log to a file if you wish to see output from the MSBuild Tasks

We are investigating the FreeBSD port selection issue.

Thanks!

**Comment by erghe** [comment]
Hi, maybe the problem is related to the "ps aux" command that is given through ssh in order to get the PID of rsync daemon. In my case, I try to run rsync in a container where ps aux is not available.

**Comment by itodirel** [comment]
I was able to reproduce the bug.

**Comment by itodirel** [comment]
We have a fix for this, with full FreeBSD support coming out in a Preview of VS on near end of June

**Comment by itodirel** [comment]
This is now fixed, and a release is out (16.7 Preview 3), please try it. We now have FreeBSD support.

**Comment by benmcmorran** [comment]
This issue is being closed because we are no longer tracking VS Linux issues on GitHub. If the issue still reproduces, bugs and new suggestions should be reported on the [Developer Community forum](https://developercommunity.visualstudio.com). It is centralized, better equipped for group voting, and has more visibility.
